mongo.go

```diff
@@ -64,14 +52,36 @@ func newItem(i *mongoItem) *resource.Item {
 	}
 }
 
+// Handler handles resource storage in a MongoDB collection.
+type Handler func(ctx context.Context) (*mgo.Collection, error)
+
+// NewHandler creates an new mongo handler
+func NewHandler(s *mgo.Session, db, collection string) Handler {
+	s.EnsureSafe(&mgo.Safe{})
+	return func(ctx context.Context) (*mgo.Collection, error) {
+		// With mgo, session.Copy() pulls a connection from the connection pool
+		return s.Copy().DB(db).C(collection), nil
+	}
+}
+
 // C returns the mongo collection managed by this storage handler
 // from a Copy() of the mgo session.
-func (m *Handler) c(ctx context.Context) (*mgo.Collection, error) {
+func (m Handler) c(ctx context.Context) (*mgo.Collection, error) {
 	if err := ctx.Err(); err != nil {
 		return nil, err
 	}
+	c, err := m(ctx)
+	if err != nil {
+		return nil, err
+	}
+	db := c.Database.Name
+
+	if ctxdbname, ok := ctx.Value(Hkey).(string); ok && ctxdbname != "" {
```

Comment:
This should stay in YOUR custom implementation. For your use-case, you would have a custom NewHandler with a function that extract the collection from the context.